src/interface.jl

```
@@ -421,6 +421,9 @@ end
 Returns a cache used with [`value_and_pullback!!`](@ref). See that function for more info.
 """
 function prepare_pullback_cache(fx...; kwargs...)
+    # Handle forward pass's primal exceptions before rule construction (zero_tangent is not defined for Ptr)
```

> Handle forward pass's primal exceptions before rule construction (zero_tangent is not defined for Ptr)

I am confused, does `prepare_pullback_cache` support `Ptr`-typed argument in `fx`? If not, let's throw an error before proceeding with the steps below.

---

Given that it currently fails for `Ptr`-typed arguments, I suggest we check for `Ptr` arguments and throw an argument error early on. 

Then, you can revert to `__exclude_unsupported_output(primal(y))` after calling `y, rvs!! = rule(map((x, dx) -> CoDual(x, fdata(dx)), fx, tangents)...)`.

---

Yes, Ptr-typed output should be forbidden too.

---

> Given that it currently fails for `Ptr`-typed arguments, I suggest we check for `Ptr` arguments and throw an argument error early on. 
> 
> Then, you can revert to `__exclude_unsupported_output(primal(y))` after calling `y, rvs!! = rule(map((x, dx) -> CoDual(x, fdata(dx)), fx, tangents)...)`. 

`__exclude_unsupported_output(primal(y))` would have to be called before the rule though as the rule call fails when an output is a ptr.

---

basically it fails at 2 places - the forward pass for the derived rules or while generate tangent data for the rule's forward pass. to avoid this im added the changes to `__exclude_unsupported_output(y)` before the rule gen itself, this way we dont need to construct a rule whos forward pass would fail.